### PR TITLE
[PYTHON API] release GIL

### PR DESCRIPTION
**Pull request by AlexeyLebedev1**
### Details:
 - *The goal is to allow another python thread work by do not holding the GIL while running c++ code, which doesn't use any python objects.*
 - *Tests check if the GIL is released safety, by incrementing reference counting of python objects while running function that releases the GIL. If function works with python object while gil is released tests fail with segfault.*
### Tickets:
 - *77646*

### CODE REVIEW

**Review by jiwaszki — COMMENTED** [review]
Discussed offline - need some detailed testing and docs notes (functions annotations, wiki page).

File: src/bindings/python/src/pyopenvino/core/core.cpp [hunk 1]
```diff
@@ -319,7 +328,9 @@ void regclass_Core(py::module m) {
             _stream << model_stream
                            .attr("read")()  // alternative: model_stream.attr("get_value")()
                            .cast<std::string>();
+            py::gil_scoped_release release;
             return self.import_model(_stream, device_name, {properties.begin(), properties.end()});
+            ;
```
Comment (jiwaszki) [inline src/bindings/python/src/pyopenvino/core/core.cpp:333]: .

File: src/bindings/python/src/pyopenvino/core/compiled_model.cpp [hunk 2]
```diff
@@ -47,7 +48,10 @@ void regclass_CompiledModel(py::module m) {
             auto request = self.create_infer_request();
             // Update inputs if there are any
             Common::set_request_tensors(request, inputs);
-            request.infer();
+            {
+                py::gil_scoped_release release;
```
Comment (jiwaszki) [inline src/bindings/python/src/pyopenvino/core/compiled_model.cpp:52]: No info on releasing GIL.

Moreover please connect it with this statement: "Blocks all methods of CompiledModel while the request is running".

File: src/bindings/python/src/pyopenvino/core/compiled_model.cpp [hunk 3]
```diff
@@ -33,6 +33,7 @@ void regclass_CompiledModel(py::module m) {
         [](ov::CompiledModel& self) {
             return std::make_shared<InferRequestWrapper>(self.create_infer_request(), self.inputs(), self.outputs());
         },
+        py::call_guard<py::gil_scoped_release>(),
```
Comment (jiwaszki) [inline src/bindings/python/src/pyopenvino/core/compiled_model.cpp:36]: No info on releasing GIL.

File: src/bindings/python/src/pyopenvino/core/core.cpp [hunk 4]
```diff
@@ -244,8 +249,11 @@ void regclass_Core(py::module m) {
 
     cls.def(
         "read_model",
-        [](ov::Core& self, py::object model, py::object weights) {
-            return self.read_model(py::str(model), py::str(weights));
+        [](ov::Core& self, py::object model_path, py::object weights_path) {
+            std::string model_path_cpp{py::str(model_path)};
+            std::string weights_path_cpp{py::str(weights_path)};
+            py::gil_scoped_release release;
```
Comment (jiwaszki) [inline src/bindings/python/src/pyopenvino/core/core.cpp:255]: No info on releasing GIL.

File: src/bindings/python/src/pyopenvino/core/core.cpp [hunk 5]
```diff
@@ -273,6 +281,7 @@ void regclass_Core(py::module m) {
             _stream << model_stream;
             return self.import_model(_stream, device_name, {properties.begin(), properties.end()});
         },
+        py::call_guard<py::gil_scoped_release>(),
```
Comment (jiwaszki) [inline src/bindings/python/src/pyopenvino/core/core.cpp:284]: No info on releasing GIL.

File: src/bindings/python/src/pyopenvino/core/core.cpp [hunk 6]
```diff
@@ -319,6 +328,7 @@ void regclass_Core(py::module m) {
             _stream << model_stream
                            .attr("read")()  // alternative: model_stream.attr("get_value")()
                            .cast<std::string>();
+            py::gil_scoped_release release;
```
Comment (jiwaszki) [inline src/bindings/python/src/pyopenvino/core/core.cpp:331]: No info on releasing GIL.

File: src/bindings/python/src/pyopenvino/core/async_infer_queue.cpp [hunk 7]
```diff
@@ -229,54 +225,45 @@ void regclass_AsyncInferQueue(py::module m) {
...
-        Waits for all InferRequests in a pool to finish scheduled work. 
+    cls.def("wait_all",
+            &AsyncInferQueue::wait_all,
+            R"(
+            One of 'flow control' functions. Blocking call.
+            Waits for all InferRequests in a pool to finish scheduled work.
 
-        Function releases GIL, other threads can work while this function waits.
-    )");
+            Function releases GIL, other threads can work while this function waits.
```
Comment (jiwaszki) [inline src/bindings/python/src/pyopenvino/core/async_infer_queue.cpp:234]:
```suggestion
            Function releases GIL, other threads can work while this function is running in the background.
```
I think this is a better overall statement. Please copy it to other docs as well.

**Review by jiwaszki — CHANGES_REQUESTED** [review]
Missing docs notes in functions that release GIL.

**Comment by akuporos** [comment]
@AlexeyLebedev1 please, update your branch with master

File: src/bindings/python/src/pyopenvino/core/infer_request.cpp [hunk 8]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (alalek) [inline src/bindings/python/src/pyopenvino/core/infer_request.cpp:1]: unrelated change. Breaks license header.

File: docs/snippets/ov_python_exclusives.py [hunk 9]
```diff
@@ -131,3 +131,39 @@ def f(request, userdata):
 
 assert all(data_done)
 #! [asyncinferqueue_set_callback]
+
+#! [releasing_gil]
+import openvino.runtime as ov
+import cv2
```
Comment (alalek) [inline docs/snippets/ov_python_exclusives.py:137]: consider using `import cv2 as cv` + `cv.` (`cv2` is going to be deprecated)

File: docs/OV_Runtime_UG/Python_API_exclusives.md [hunk 10]
```diff
@@ -141,3 +141,18 @@ The callback of `AsyncInferQueue` is uniform for every job. When executed, GIL i
...
+
+@sphinxdirective
+
+.. doxygensnippet:: docs/snippets/ov_python_exclusives.py
+    :language: python
+    :fragment: [releasing_gil]
+
+@endsphinxdirective
+
+> **NOTE**: While GIL is released functions can still modify and/or operate on Python objects in C++, thus there is no reference counting. User is responsible for thread safety if sharing of these objects with other thread occurs. It can affects your code only if multiple threads are spawned in Python.
```
Comment (alalek) [inline docs/OV_Runtime_UG/Python_API_exclusives.md:158]: > While GIL is released ... operate on Python objects

- we can modify C++ state of OpenVINO objects - this doesn't break Python state protected by GIL.
- &#x2757; we can't query or modify state of any Python objects
- &#x2757; also we can't call "Python" callbacks (reacquire GIL before)

so this note is not accurate.

Comment (AlexeyLebedev1) [reply docs/OV_Runtime_UG/Python_API_exclusives.md:158]: For sure python objects in this meaning are not modified, but GIL is only protects reference counters in this case. But what if  C++ state of python object is modifying simultaneously in several python threads? It can be not thread safety and GIL doesn't protect it.

File: src/bindings/python/src/pyopenvino/core/infer_request.cpp (hunk 8 above)
Comment (AlexeyLebedev1) [reply src/bindings/python/src/pyopenvino/core/infer_request.cpp]: Removed, thanks.

File: docs/snippets/ov_python_exclusives.py (hunk 9 above)
Comment (AlexeyLebedev1) [reply docs/snippets/ov_python_exclusives.py]: Done.

File: src/bindings/python/src/pyopenvino/core/compiled_model.cpp (hunk 3 above)
Comment (AlexeyLebedev1) [reply src/bindings/python/src/pyopenvino/core/compiled_model.cpp:36]: Updated.

File: src/bindings/python/src/pyopenvino/core/core.cpp (hunk 4 above)
Comment (AlexeyLebedev1) [reply src/bindings/python/src/pyopenvino/core/core.cpp:269]: Updated.

File: src/bindings/python/src/pyopenvino/core/core.cpp (hunk 6 above)
Comment (AlexeyLebedev1) [reply src/bindings/python/src/pyopenvino/core/core.cpp:349]: Updated.

File: src/bindings/python/src/pyopenvino/core/async_infer_queue.cpp (hunk 7 above)
Comment (AlexeyLebedev1) [reply src/bindings/python/src/pyopenvino/core/async_infer_queue.cpp]: Updated.
